src/sentry/auth/access.py

```diff
@@ -100,45 +102,83 @@ class Access(abc.ABC):
-                organizationmember=self.member, is_active=True, team__status=TeamStatus.VISIBLE
+                organizationmember=self._member, is_active=True, team__status=TeamStatus.VISIBLE
             ).select_related("team")
         }
 
     @cached_property
-    def teams(self) -> FrozenSet[Team]:
-        """Return the set of teams in which the user has actual membership."""
-        return frozenset(self._team_memberships.keys())
+    def team_ids_with_membership(self) -> FrozenSet[int]:
+        """Return the IDs of teams in which the user has actual membership.
+
+        This represents the set of all teams for which `has_team_membership` returns
+        true. Use that method where possible and use this property only when you need
+        to iterate or query for all such teams.
+
+        Compare to accessible_team_ids, which is equal to this property in the
+        typical case but represents a superset of IDs in case of superuser access.
+        """
+        return frozenset(team.id for team in self._team_memberships.keys())
+
+    @property
+    def accessible_team_ids(self) -> FrozenSet[int]:
+        """Return the IDs of teams to which the user has access.
+
+        This represents the set of all teams for which `has_team_access` returns
+        true. Use that method where possible and use this property only when you need
+        to iterate or query for all such teams.
+        """
+        return self.team_ids_with_membership
 
     @cached_property
-    def projects(self) -> FrozenSet[Project]:
-        """Return the set of projects to which the user has access via actual team membership."""
+    def project_ids_with_team_membership(self) -> FrozenSet[int]:
+        """Return the IDs of projects to which the user has access via actual team membership.
 
-        teams = self.teams
+        This represents the set of all projects for which `has_project_membership`
+        returns true. Use that method where possible and use this property only when
+        you need to iterate or query for all such teams.
+
+        Compare to accessible_project_ids, which is equal to this property in the
+        typical case but represents a superset of IDs in case of superuser access.
+        """
+
+        teams = self._team_memberships.keys()
         if not teams:
             return frozenset()
 
         with sentry_sdk.start_span(op="get_project_access_in_teams") as span:
             projects = frozenset(
-                Project.objects.filter(status=ProjectStatus.VISIBLE, teams__in=teams).distinct()
+                Project.objects.filter(status=ProjectStatus.VISIBLE, teams__in=teams)
+                .distinct()
+                .values_list("id", flat=True)
             )
             span.set_data("Project Count", len(projects))
             span.set_data("Team Count", len(teams))
 
         return projects
 
+    @property
+    def accessible_project_ids(self) -> FrozenSet[int]:
```

Comment:
Same with `membership_projects` and `accessible_projects` for more consistent verb,noun ordering 🤷